source/MaterialXCore/Document.cpp

```diff
@@ -182,47 +182,83 @@ void Document::initialize()
     setVersionIntegers(MATERIALX_MAJOR_VERSION, MATERIALX_MINOR_VERSION);
 }
 
-NodeDefPtr Document::addNodeDefFromGraph(const NodeGraphPtr nodeGraph, const string& nodeDefName, const string& node,
-                                         const string& version, bool isDefaultVersion, const string& group, const string& newGraphName)
+NodeDefPtr Document::addNodeDefFromGraph(const NodeGraphPtr nodeGraph, const string& nodeDefName,
+                                         const string& category, const string& newGraphName)
 {
+    if (category.empty())
+    {
+        throw Exception("Cannote create a nodedef without a category identifier");
+    }
+
     if (getNodeDef(nodeDefName))
     {
         throw Exception("Cannot create duplicate nodedef: " + nodeDefName);
     }
 
-    NodeGraphPtr graph = nodeGraph;
-    if (!newGraphName.empty())
+    if (getNodeGraph(newGraphName))
     {
-        if (getNodeGraph(newGraphName))
-        {
-            throw Exception("Cannot create duplicate nodegraph: " + newGraphName);
-        }
-        graph = addNodeGraph(newGraphName);
-        graph->copyContentFrom(nodeGraph);
+        throw Exception("Cannot create duplicate nodegraph: " + newGraphName);
     }
+    
+    // Create a new functional nodegraph, and copy over the
+    // contents from the compound nodegraph
+    NodeGraphPtr graph = addNodeGraph(newGraphName);
+    graph->copyContentFrom(nodeGraph);
+
+    for (auto graphChild : graph->getChildren())
+    {
+        graphChild->removeAttribute("xpos");
+        graphChild->removeAttribute("ypos");
+    }        
     graph->setNodeDefString(nodeDefName);
 
+    // Create a new nodedef and set it's category
     NodeDefPtr nodeDef = addChild<NodeDef>(nodeDefName);
-    nodeDef->setNodeString(node);
-    if (!group.empty())
-    {
-        nodeDef->setNodeGroup(group);
-    }
+    nodeDef->setNodeString(category);
 
-    if (!version.empty())
-    {
-        nodeDef->setVersionString(version);
+    // Expose any existing interfaces from the graph.
+    // Any connection attributes ("nodegraph", "nodename", "channels", "interfacename") on the 
+    // existing interface should be removed from the definition as well as any source URI.
+
+    // Attributes which should not be copied over
+    StringSet filterAttributes = { PortElement::NODE_GRAPH_ATTRIBUTE, PortElement::NODE_NAME_ATTRIBUTE, 
+                                   PortElement::CHANNELS_ATTRIBUTE, PortElement::INTERFACE_NAME_ATTRIBUTE,
+                                   "xpos", "ypos" };
```

Comment:
It's probably time to add these two strings as static members of the `Element` class, so that we don't continue propagating the string literals throughout the codebase.  What are your thoughts?

Reply:
Agreed. We should be using predefined constant strings especially if temp string allocation can affect performance. I'll add the constants and replace all existing instances.